tests/unit/jobserver/models/test_paired_fields.py

```diff
@@ -0,0 +1,301 @@
+"""
+
+# Several tests rely on models having factories. We explicitly ignore those that don't.
+MODELS_WITHOUT_FACTORIES = [
+    ("applications.CmoPriorityListPage",),
+    ("applications.LegalBasisPage",),
+    ("jobserver.ProjectMembership",),
+]
+
+
+@pytest.mark.parametrize("model", utils.get_models())
+@pytest.mark.parametrize("lsuffix,rsuffix", [("_at", "_by")])
+def test_fields_are_paired(model, lsuffix, rsuffix):
+    """
+    Many fields look like they are paired, but (sadly) aren't. Rather than maintain a
+    list of paired fields (includes), which is error-prone, we maintain an list of
+    unpaired fields (excludes).
+
+    If a new field is added, and the new field looks like it is paired, then this test
+    will fail. Either add another new field or exclude the new field.
+    """
+    names = [
+        name
+        for field in model._meta.fields
+        if (name := field.name)
+        not in MODEL_TO_UNPAIRED_FIELDS.get(model._meta.label, set())
+    ]
+
+    def get_stems(suffix):
+        return {stem for name in names if (stem := name.removesuffix(suffix)) != name}
+
+    lstems = get_stems(lsuffix)
+    rstems = get_stems(rsuffix)
+
+    unpaired_lnames = [f"{stem}{lsuffix}" for stem in lstems - rstems]
+    unpaired_rnames = [f"{stem}{rsuffix}" for stem in rstems - lstems]
+    unpaired_names = sorted(unpaired_lnames + unpaired_rnames)
+
+    assert (
+        not unpaired_names
+    ), f"{model._meta.label} is missing paired fields for: {', '.join(unpaired_names)}"
+
+
+@pytest.mark.parametrize(
+    "factory,at_name,by_name",
+    utils.get_all_paired_fields(
+        "_at",
+        "_by",
+        exclude=[
+            ("jobserver.PublishRequest", "decision_at", "decision_by", "decision"),
+            # Missing ReleaseFileReviewFactory.comments
+            ("jobserver.ReleaseFileReview",),
+            # Missing RedirectFactory.analysis_request
+            ("redirects.Redirect",),
+        ]
+        + MODELS_WITHOUT_FACTORIES,
+    ),
+)
+def test_both_fields_set(factory, at_name, by_name):
+    """
+    Test the Value/Value case (success).
+    """
+    factory(**{at_name: timezone.now(), by_name: UserFactory()})
+
+
+@pytest.mark.parametrize(
+    "factory,at_name,by_name",
+    utils.get_optional_paired_fields(
+        "_at",
+        "_by",
+        exclude=[
+            # Missing RedirectFactory.analysis_request
+            ("redirects.Redirect",),
+        ]
+        + MODELS_WITHOUT_FACTORIES,
+    ),
+)
+def test_neither_field_set(factory, at_name, by_name):
+    """
+    Test the None/None case (success), only when both fields are optional.
+    """
+    factory(**{at_name: None, by_name: None})
+
+
+@pytest.mark.django_db(transaction=True)
+@pytest.mark.parametrize(
+    "factory,at_name,by_name",
+    utils.get_all_paired_fields(
+        "_at",
+        "_by",
+        exclude=[
+            # "at" has auto_now=True
```

Comment:
That's both useful and interesting, thank you.

I think excluding the models where the "at" field has `auto_now=True` is desirable. (In other words, I don't think it's desirable to modify the test to accommodate these models.) That's because this test, and its three siblings, test the create path, and not the update path. Possibly, they should test both. As you suggest, though, that's for another day 🙂.

Reply:
All excellent points. Thank you :)